op-program/client/l2/canon_test.go

```diff
@@ -0,0 +1,108 @@
+package l2
+
+import (
+	"testing"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/stretchr/testify/require"
+)
+
+func TestCanonicalBlockNumberOracle_GetHeaderByNumber(t *testing.T) {
+	headBlockNumber := 3
+	blockCount := 3
+	chainCfg, blocks, oracle := setupOracle(t, blockCount, headBlockNumber, true)
+	head := blocks[headBlockNumber].Header()
+
+	blockByHash := func(hash common.Hash) *types.Block {
+		return oracle.BlockByHash(hash, chainCfg.ChainID.Uint64())
+	}
+	canon := NewCanonicalBlockHeaderOracle(head, blockByHash)
+	require.Equal(t, head.Hash(), canon.CurrentHeader().Hash())
+	require.Nil(t, canon.GetHeaderByNumber(4))
+
+	oracle.Blocks[blocks[3].Hash()] = blocks[3]
+	h := canon.GetHeaderByNumber(3)
+	require.Equal(t, blocks[3].Hash(), h.Hash())
+
+	oracle.Blocks[blocks[2].Hash()] = blocks[2]
+	h = canon.GetHeaderByNumber(2)
+	require.Equal(t, blocks[2].Hash(), h.Hash())
+
+	oracle.Blocks[blocks[1].Hash()] = blocks[1]
+	h = canon.GetHeaderByNumber(1)
+	require.Equal(t, blocks[1].Hash(), h.Hash())
+
+	oracle.Blocks[blocks[0].Hash()] = blocks[0]
+	h = canon.GetHeaderByNumber(0)
+	require.Equal(t, blocks[0].Hash(), h.Hash())
+
+	// Test eraliest block short-circuiting. Do not expect oracle requests for other blocks.
+	oracle.Blocks = map[common.Hash]*types.Block{
+		blocks[1].Hash(): blocks[1],
+	}
+	require.Equal(t, blocks[1].Hash(), canon.GetHeaderByNumber(1).Hash())
```

Comment:
This is actually testing that we use the block hash cache.  And given the above change we now need to allow requesting block 1 again.
```suggestion
	// Test that the block hash is cached. Do not expect oracle requests for any other blocks.
	// Allow requesting block 1 again as we're specifically asking for it and only the hash is cached
	requestedBlocks[blocks[1].Hash()] = false
	oracle.Blocks = map[common.Hash]*types.Block{
		blocks[1].Hash(): blocks[1],
	}
	require.Equal(t, blocks[1].Hash(), canon.GetHeaderByNumber(1).Hash())
```